// eslint-plugin-react-native-a11y/lib/rules/has-valid-accessibility-state.js
"use strict";

var _jsxAstUtils = require("jsx-ast-utils");

var _schemas = require("../util/schemas");

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// ----------------------------------------------------------------------------
// Rule Definition
// ----------------------------------------------------------------------------
var validKeys = ['disabled', 'selected', 'checked', 'busy', 'expanded'];
module.exports = {
  meta: {
    docs: {},
    schema: [(0, _schemas.generateObjSchema)()]
  },
  create: function create(context) {
    return {
      JSXAttribute: function JSXAttribute(node) {
        var attrName = (0, _jsxAstUtils.elementType)(node);

        if (attrName === 'accessibilityState') {
          var attrValue = (0, _jsxAstUtils.getPropValue)(node);

          var error = function error(message) {
            return context.report({
              node,
              message
            });
          };

          if (typeof attrValue !== 'object' || Array.isArray(attrValue)) {
            error('accessibilityState must be an object');
          } else {
            Object.entries(attrValue).map(function (_ref) {
              var _ref2 = _slicedToArray(_ref, 2),
                  key = _ref2[0],
                  value = _ref2[1];

              // $FlowFixMe
              var astObjectProp = node.value.expression.properties.find( // $FlowFixMe
              function (f) {
                return f.key.name === key;
              }); // we can't determine the associated value type of an Identifier expression
              // treat these cases as though they are valid
              // $FlowFixMe

              if (astObjectProp && astObjectProp.value.type !== 'Identifier') {
                if (validKeys.indexOf(key) < 0) {
                  error(`accessibilityState object: "${key}" is not a valid key`);
                } else if (key === 'checked' && !(typeof value === 'boolean' || value === 'mixed')) {
                  error(`accessibilityState object: "checked" value is not either a boolean or 'mixed'`);
                } else if (key !== 'checked' && typeof value !== 'boolean') {
                  error(`accessibilityState object: "${key}" value is not a boolean`);
                }
              }
            });
          }
        }
      }
    };
  }
};
